### PR TITLE
AMP-48: School CRUD Page

### PR DESCRIPTION
**Pull request by ksdfg**

### CODE REVIEW

**Comment by akhilnarang** [comment]
I know its still a draft, but just thought I'd note down whatever I faced

1. Enter details to create a new school, hit enter
2. Loader pops up, school gets created, modal still stays, with the values I just entered

I guess we should either clear the text, or dismiss it after the loader
And error handling, which I guess is planned anyway.

Faced this in the toast sometimes -
image

I also managed to break CORS along the way :)

**Comment by ksdfg** [comment]
> I know its still a draft, but just thought I'd note down whatever I faced
> 
>     1. Enter details to create a new school, hit enter
> 
>     2. Loader pops up, school gets created, modal still stays, with the values I just entered
> 
> 
> I guess we should either clear the text, or dismiss it after the loader
> And error handling, which I guess is planned anyway.
> 
> Faced this in the toast sometimes -
> image
> 
> I also managed to break CORS along the way :)

Fixed the toast, silly mistake on my end.

As for the modal closing after operation is done, I actually wanted to try and do that for a while, but I couldn't figure out how to. I don't plan on adding any specific validation as of now, you guys can hopefully do that when you have time.

File: src/pages/schools/components/SchoolModal.jsx [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            } else {
+                toast.error("Please update some information 😓", {
+                    position: toast.POSITION.TOP_CENTER,
+                });
+            }
+        } else {
+            if (submit.hasOwnProperty("name") && submit["name"]) {
+                this.props.CreateSchool(submit);
+            } else {
+                toast.error("Please add email, type and password 😓", {
```
Comment (PranavBakre) [inline src/pages/schools/components/SchoolModal.jsx]: What is this for?

Comment (ksdfg) [reply src/pages/schools/components/SchoolModal.jsx]: hmm, copy pasting is never a good idea

Comment (PranavBakre) [reply src/pages/schools/components/SchoolModal.jsx]: also why hasOwnProperty?

Comment (ksdfg) [reply src/pages/schools/components/SchoolModal.jsx]: that entire part is now redundant tbh, since I'm validating all inputs now anyway. This was just a final check before submitting to make sure there is some form of validation available

**Comment by PranavBakre** [comment]
> > I know its still a draft, but just thought I'd note down whatever I faced
> > ```
> > 1. Enter details to create a new school, hit enter
> > 
> > 2. Loader pops up, school gets created, modal still stays, with the values I just entered
> > ```
> > 
> > 
> > I guess we should either clear the text, or dismiss it after the loader
> > And error handling, which I guess is planned anyway.
> > Faced this in the toast sometimes -
> > image
> > I also managed to break CORS along the way :)
> 
> Fixed the toast, silly mistake on my end.
> 
> As for the modal closing after operation is done, I actually wanted to try and do that for a while, but I couldn't figure out how to. I don't plan on adding any specific validation as of now, you guys can hopefully do that when you have time.




you can use onClose directly in componentDidUpdate na? right under where toast.success is called

**Comment by PranavBakre** [comment]
@PranavBakre ! ignore this. markdown reeped this before I had to update

**Comment by ksdfg** [comment]
> > > I know its still a draft, but just thought I'd note down whatever I faced
> > > ```
> > > 1. Enter details to create a new school, hit enter
> > > 
> > > 2. Loader pops up, school gets created, modal still stays, with the values I just entered
> > > ```
> > > 
> > > 
> > > I guess we should either clear the text, or dismiss it after the loader
> > > And error handling, which I guess is planned anyway.
> > > Faced this in the toast sometimes -
> > > image
> > > I also managed to break CORS along the way :)
> > 
> > 
> > Fixed the toast, silly mistake on my end.
> > As for the modal closing after operation is done, I actually wanted to try and do that for a while, but I couldn't figure out how to. I don't plan on adding any specific validation as of now, you guys can hopefully do that when you have time.
> 
> you can use onClose directly in componentDidUpdate na? right under where toast.success is called

my bad, I forgot to update here, the modal now closes on update/create.
was done during the rush before the demo, hence forgot to update comment.

**Comment by PranavBakre** [comment]
@ksdfg but why did u use componentDidUpdate when you had an closeModal

**Comment by ksdfg** [comment]
@PranavBakre so as to not have modal state var as open in index.jsx iirc

File: src/pages/dashboard/index.jsx [hunk 2]
```diff
@@ -36,10 +37,13 @@ class Dashboard extends Component {
     }
 
     componentDidUpdate(prevProps, prevState, snapshot) {
-        if (prevProps.errorMessage !== this.props.errorMessage && this.props.errorMessage !== "") {
-            toast.error(`Error 😓: ${this.props.errorMessage}`, {
-                position: toast.POSITION.TOP_CENTER,
-            });
+        // check if any new error message needs to be displayed
+        for (let i = 0; i < this.props.errorMessage.length; i++) {
```
Comment (PranavBakre) [inline src/pages/dashboard/index.jsx]: Could you try array.forEach loop here. Checking with length looks weird when you have functions that can iterate. Array.foreach can also get you index.

File: src/redux/reducer.js [hunk 3]
```diff
@@ -14,12 +15,13 @@ const persistConfig = {
...
     auth: AuthReducer,
     user: UserReducer,
+    school: SchoolReducer,
 });
 
 const RootReducer = (state, action) => {
     if (action.type === AuthActionTypes.LOGOUT) {
         storage.removeItem("persist:auth");
-        state.auth = state.user = undefined;
+        state.auth = state.user = state.school = undefined;
```
Comment (PranavBakre) [inline src/redux/reducer.js:24]: A thought. What if you retrieve _persist and clear state and set it back? This wouldnt require changes over time when new modules are added

https://github.com/VPPR/frontend/blob/9d9a2463a6f2b72a7be0aa62f995f7a0467f2077/src/redux/reducer.js#L25

Comment (ksdfg) [reply src/redux/reducer.js:24]: that's probably a good idea. Might make a `misc` PR for small changes like this one....

File: src/pages/dashboard/index.jsx (hunk 2 above)
Comment (ksdfg) [reply src/pages/dashboard/index.jsx]: I did try with forEach before, but there was some issue iirc
let me try again now that I have slept a little

Comment (ksdfg) [reply src/pages/dashboard/index.jsx]: k nvm, found a huge bug with this

If we trigger the same error multiple times, it only shows up the first time

Comment (ksdfg) [reply src/pages/dashboard/index.jsx]: because technically the error did not change

Comment (PranavBakre) [reply src/pages/dashboard/index.jsx]: wouldnt this happen right now as well?

Comment (PranavBakre) [reply src/pages/dashboard/index.jsx]: Also why would multiple same errors not work? If I am right, every action that could leads to an error first sets errorMessage to ""

Comment (ksdfg) [reply src/pages/dashboard/index.jsx]: I mean, I just tried, didn't get any errors

and yeah, that is happening right now as well

Comment (ksdfg) [reply src/pages/dashboard/index.jsx]: now I'm not sure if it's because of my code or the damn cors issue

Comment (ksdfg) [reply src/pages/dashboard/index.jsx]: hmmm, it seems to be working for create school and create user errors, so maybe this can be chalked up to no errors being set by the CORS errors...

**Review by akhilnarang — APPROVED** [review]
Tested basic functionality, seems good to me!
